### PR TITLE
App: Demonstrate lidar sensor streaming with HoloViz

### PR DESCRIPTION
**Pull request by tbirdso**
Adds an operator and application to demonstrate streaming from the Velodyne VLP-16 3D lidar sensor to display with HoloViz.

Change:
- Implements C++ `VelodyneLidarOp` for processing VLP-16 packets on GPU, informed by the NVIDIA DeepMap SDK
- Implements C++ `VelodyneLidarApp` demonstration pipeline to receive VLP-16 packets over IP/UDP, process them into a Cartesian point cloud on GPU, and then pass them to HoloViz for display.

Portions of CUDA VLP-16 processing code are contributed by the NVIDIA nvMap team via the NVIDIA DeepMap SDK. Collaborators are credited in metadata files.

### CODE REVIEW

**Review by jjomier — APPROVED** [review]
LGTM!

**Comment by tbirdso** [comment]
Thanks @jjomier! Will leave this PR open a little longer to allow others to review.

File: operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  size_t size_of_pitch_table = kVLP16LineCount * sizeof(double);
+  vlp16_sin_pitch_ = (double*)malloc(size_of_pitch_table);
+  vlp16_cos_pitch_ = (double*)malloc(size_of_pitch_table);
+
+  for (int i = 0; i < kVLP16LineCount; i++) {
+    vlp16_sin_pitch_[i] = sin(data_collection::sensors::kVLP16PitchDegrees[i] * kDegreesToRadians);
+    vlp16_cos_pitch_[i] = cos(data_collection::sensors::kVLP16PitchDegrees[i] * kDegreesToRadians);
+  }
+
+  CHECK_CUDA_ERROR(cudaMalloc((void**)&d_sin_rot_table_, size_of_rot_table));
```
Comment (AndreasHeumann) [inline operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu:113]: It seems all the CUDA memory allocated here is never freed.

File: operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  d_xyzintensity_list->blocks_[j].lines_[i].y =
+      range_meters * d_vlp16_cos_pitch[i] * d_cos_rot_table[azimuth];
+  d_xyzintensity_list->blocks_[j].lines_[i].z = range_meters * d_vlp16_sin_pitch[i];
+}
+
+/// @brief Pre-compute sine and cosine values for rapid lookup.
+void VelodyneConvertXYZHelper::InitSinAndCosTable() {
+  HOLOSCAN_LOG_DEBUG("Initializing sin and cos tables in GPU.");
+  size_t size_of_rot_table = kVelodyneMaxAzimuth * sizeof(double);
+  sin_rot_table_ = (double*)malloc(size_of_rot_table);
```
Comment (AndreasHeumann) [inline operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu:96]: The CPU memory for the sin and cos tables is not used outside of this function. You might want to have a local variable for this. Also using `std::unique_ptr` should simplify things.

File: operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+#include <holoscan/holoscan.hpp>
+
+// Defines all tables.
+// Use constant memory cause it's 100X faster than shared memory.
+__constant__ double d_vlp16_sin_pitch[data_collection::sensors::kVLP16LineCount];
+__constant__ double d_vlp16_cos_pitch[data_collection::sensors::kVLP16LineCount];
+
+// Quick return cuda error and error message. Will force quit program cause all
+// cuda errors are fatal errors.
+#define CHECK_CUDA_ERROR(err) (HandleError(err, __FILE__, __LINE__))
```
Comment (AndreasHeumann) [inline operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu:35]: Could you use the same CUDA_TRY macro as in the other file below for consistency?

File: operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            file,
+            line);
+    exit(int(err));
+  }
+}
+
+namespace data_collection {
+namespace sensors {
+
+__device__ uint16_t ConvertSmallEndianToBigEndian(uint16_t small_endian) {
```
Comment (AndreasHeumann) [inline operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu:52]: `SmallEndian` is new to me, should this be `LittleEndian`?

**Review by AndreasHeumann — COMMENTED** [review]
Nice! I added some comments.

File: operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu (hunk 2 above)
Comment (tbirdso) [reply operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu]: Thanks, good catch. Fixed with local `std::vector` for array memory management.

File: operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu (hunk 3 above)
Comment (tbirdso) [reply operators/velodyne_lidar/cpp/velodyne_convert_xyz.cu]: Fixed

**Comment by tbirdso** [comment]
Thanks @AndreasHeumann ! Updated to address your review comments.

**Pull request merged**